File: sdk/python/aragora_sdk/namespaces/external_agents.py
"""
External Agents Namespace API

Provides methods for external agent operations:
- Adapter discovery and health monitoring
- Task submission, tracking, and cancellation
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


class ExternalAgentsAPI:
    """
    Synchronous External Agents API.

    Provides methods for managing external agent adapters and tasks:
    - List available adapters
    - Check adapter health
    - Submit, track, and cancel tasks

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai")
        >>> adapters = client.external_agents.list_adapters()
        >>> task = client.external_agents.submit_task(
        ...     task_type="code_review",
        ...     prompt="Review this pull request for security issues",
        ... )
    """

    def __init__(self, client: AragoraClient):
        self._client = client

    # ===========================================================================
    # Adapter Discovery
    # ===========================================================================

    def list_adapters(self) -> dict[str, Any]:
        """
        List available external agent adapters.

        Returns:
            Dict with adapters array and their configurations
        """
        return self._client.request("GET", "/api/v1/external-agents/adapters")

    def get_adapter_health(
        self,
        adapter_name: str | None = None,
    ) -> dict[str, Any]:
        """
        Get health status for external agent adapters.

        Args:
            adapter_name: Filter by specific adapter name (optional)

        Returns:
            Dict with adapter health status information
        """
        params: dict[str, Any] = {}
        if adapter_name:
            params["adapter"] = adapter_name
        return self._client.request("GET", "/api/v1/external-agents/health", params=params)

    # ===========================================================================
    # Task Management
    # ===========================================================================

    def submit_task(
        self,
        task_type: str,
        prompt: str,
        adapter: str = "openhands",
        tool_permissions: list[str] | None = None,
        timeout_seconds: float = 3600.0,
        max_steps: int = 100,
        context: dict[str, Any] | None = None,
        workspace_id: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """
        Submit a task to an external agent.

        Args:
            task_type: Type of task to execute
            prompt: The prompt or instructions for the agent
            adapter: Adapter to use (default: "openhands")
            tool_permissions: List of tool permissions granted to the agent
            timeout_seconds: Maximum execution time in seconds (default: 3600.0)
            max_steps: Maximum number of steps the agent can take (default: 100)
            context: Additional context for the task
            workspace_id: Workspace to execute the task in
            metadata: Additional metadata

        Returns:
            Dict with task_id and submission status
        """
        data: dict[str, Any] = {
            "task_type": task_type,
            "prompt": prompt,
            "adapter": adapter,
            "timeout_seconds": timeout_seconds,
            "max_steps": max_steps,
            "metadata": metadata or {},
        }
        if tool_permissions is not None:
            data["tool_permissions"] = tool_permissions
        if context is not None:
            data["context"] = context
        if workspace_id is not None:
            data["workspace_id"] = workspace_id
        return self._client.request("POST", "/api/v1/external-agents/tasks", json=data)

    def get_task(self, task_id: str) -> dict[str, Any]:
        """
        Get details for a specific external agent task.

        Args:
            task_id: Task ID

        Returns:
            Dict with task details including status, result, and metadata
        """
        return self._client.request("GET", f"/api/v1/external-agents/tasks/{task_id}")

    def cancel_task(self, task_id: str) -> dict[str, Any]:
        """
        Cancel a running external agent task.

        Args:
            task_id: Task ID

        Returns:
            Dict with cancellation status
        """
        return self._client.request("DELETE", f"/api/v1/external-agents/tasks/{task_id}")


class AsyncExternalAgentsAPI:
    """
    Asynchronous External Agents API.

    Example:
        >>> async with AragoraAsyncClient(base_url="https://api.aragora.ai") as client:
        ...     adapters = await client.external_agents.list_adapters()
        ...     task = await client.external_agents.submit_task(
        ...         task_type="code_review",
        ...         prompt="Review this pull request for security issues",
        ...     )
    """

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    # ===========================================================================
    # Adapter Discovery
    # ===========================================================================

    async def list_adapters(self) -> dict[str, Any]:
        """List available external agent adapters."""
        return await self._client.request("GET", "/api/v1/external-agents/adapters")

    async def get_adapter_health(
        self,
        adapter_name: str | None = None,
    ) -> dict[str, Any]:
        """Get health status for external agent adapters."""
        params: dict[str, Any] = {}
        if adapter_name:
            params["adapter"] = adapter_name
        return await self._client.request("GET", "/api/v1/external-agents/health", params=params)

    # ===========================================================================
    # Task Management
    # ===========================================================================

    async def submit_task(
        self,
        task_type: str,
        prompt: str,
        adapter: str = "openhands",
        tool_permissions: list[str] | None = None,
        timeout_seconds: float = 3600.0,
        max_steps: int = 100,
        context: dict[str, Any] | None = None,
        workspace_id: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Submit a task to an external agent."""
        data: dict[str, Any] = {
            "task_type": task_type,
            "prompt": prompt,
            "adapter": adapter,
            "timeout_seconds": timeout_seconds,
            "max_steps": max_steps,
            "metadata": metadata or {},
        }
        if tool_permissions is not None:
            data["tool_permissions"] = tool_permissions
        if context is not None:
            data["context"] = context
        if workspace_id is not None:
            data["workspace_id"] = workspace_id
        return await self._client.request("POST", "/api/v1/external-agents/tasks", json=data)

    async def get_task(self, task_id: str) -> dict[str, Any]:
        """Get external agent task details."""
        return await self._client.request("GET", f"/api/v1/external-agents/tasks/{task_id}")

    async def cancel_task(self, task_id: str) -> dict[str, Any]:
        """Cancel a running external agent task."""
        return await self._client.request("DELETE", f"/api/v1/external-agents/tasks/{task_id}")

File: sdk/typescript/src/namespaces/external-agents.ts
/**
 * External Agents API
 *
 * Submit tasks to external AI agent frameworks (OpenHands, AutoGPT, CrewAI)
 * through Aragora's enterprise security gateway.
 */

/**
 * External adapter descriptor.
 */
export interface ExternalAdapter {
  name: string;
  description: string;
  config_class: string;
}

/**
 * Response from listing available adapters.
 */
export interface AdapterListResponse {
  adapters: ExternalAdapter[];
  total: number;
}

/**
 * Health status for a single adapter.
 */
export interface AdapterHealth {
  adapter_name: string;
  healthy: boolean;
  last_check: string;
  response_time_ms: number;
  error?: string;
  framework_version?: string;
}

/**
 * Response from the health endpoint.
 */
export interface HealthResponse {
  health: AdapterHealth[];
  total: number;
}

/**
 * Options for submitting a task to an external agent.
 */
export interface TaskSubmitOptions {
  adapter?: string;
  toolPermissions?: string[];
  timeoutSeconds?: number;
  maxSteps?: number;
  context?: Record<string, unknown>;
  workspaceId?: string;
  metadata?: Record<string, unknown>;
}

/**
 * Response after submitting a task.
 */
export interface TaskSubmitResponse {
  task_id: string;
  status: string;
  adapter: string;
}

/**
 * Information about a submitted task.
 */
export interface TaskInfo {
  task_id: string;
  status: string;
  result?: Record<string, unknown>;
}

/**
 * Response after cancelling a task.
 */
export interface CancelResponse {
  task_id: string;
  cancelled: boolean;
}

/**
 * Client interface for external agents operations.
 */
interface ExternalAgentsClientInterface {
  request<T = unknown>(
    method: string,
    path: string,
    options?: { params?: Record<string, unknown>; json?: Record<string, unknown> }
  ): Promise<T>;
}

/**
 * External Agents API namespace.
 *
 * Provides methods for submitting tasks to external AI agent frameworks
 * (OpenHands, AutoGPT, CrewAI) through Aragora's enterprise security gateway:
 * - Listing available adapters
 * - Checking adapter health
 * - Submitting, polling, and cancelling tasks
 *
 * @example
 * ```typescript
 * const client = createClient({ baseUrl: 'https://api.aragora.ai', apiKey: 'your-key' });
 *
 * // List available adapters
 * const { adapters } = await client.externalAgents.listAdapters();
 *
 * // Check adapter health
 * const { health } = await client.externalAgents.getHealth();
 *
 * // Submit a task
 * const task = await client.externalAgents.submitTask('code_review', 'Review auth module', {
 *   adapter: 'openhands',
 *   timeoutSeconds: 300,
 * });
 *
 * // Poll for results
 * const info = await client.externalAgents.getTask(task.task_id);
 *
 * // Cancel a running task
 * const result = await client.externalAgents.cancelTask(task.task_id);
 * ```
 */
export class ExternalAgentsAPI {
  constructor(private client: ExternalAgentsClientInterface) {}

  /**
   * List all registered external agent adapters.
   *
   * Returns the set of adapters available for task submission,
   * including their names, descriptions, and configuration classes.
   */
  async listAdapters(): Promise<AdapterListResponse> {
    return this.client.request(
      'GET',
      '/api/v1/external-agents/adapters',
    ) as Promise<AdapterListResponse>;
  }

  /**
   * Get health status of external agent adapters.
   *
   * Returns health information including connectivity, response time,
   * and framework version for each adapter. Optionally filter by adapter name.
   *
   * @param adapterName - Optional adapter name to check health for a specific adapter.
   */
  async getHealth(adapterName?: string): Promise<HealthResponse> {
    const params: Record<string, unknown> = {};
    if (adapterName !== undefined) {
      params.adapter = adapterName;
    }
    return this.client.request(
      'GET',
      '/api/v1/external-agents/health',
      { params },
    ) as Promise<HealthResponse>;
  }

  /**
   * Submit a task to an external agent framework.
   *
   * Sends the task through Aragora's security gateway, which enforces
   * tool permissions, timeout limits, and workspace isolation before
   * delegating to the target adapter.
   *
   * @param taskType - The type of task to execute (e.g. 'code_review', 'analysis').
   * @param prompt - The natural-language prompt describing the task.
   * @param options - Optional configuration including adapter selection,
   *   tool permissions, timeout, max steps, context, workspace, and metadata.
   */
  async submitTask(
    taskType: string,
    prompt: string,
    options?: TaskSubmitOptions,
  ): Promise<TaskSubmitResponse> {
    return this.client.request(
      'POST',
      '/api/v1/external-agents/tasks',
      {
        json: {
          task_type: taskType,
          prompt,
          adapter: options?.adapter,
          tool_permissions: options?.toolPermissions,
          timeout_seconds: options?.timeoutSeconds,
          max_steps: options?.maxSteps,
          context: options?.context,
          workspace_id: options?.workspaceId,
          metadata: options?.metadata,
        },
      },
    );
  }

  /**
   * Get the current status and result of a submitted task.
   *
   * @param taskId - The ID of the task to retrieve.
   */
  async getTask(taskId: string): Promise<TaskInfo> {
    return this.client.request(
      'GET',
      `/api/v1/external-agents/tasks/${taskId}`,
    ) as Promise<TaskInfo>;
  }

  /**
   * Cancel a running external agent task.
   *
   * Sends a cancellation request to the adapter. The task may not
   * stop immediately depending on the external framework's behavior.
   *
   * @param taskId - The ID of the task to cancel.
   */
  async cancelTask(taskId: string): Promise<CancelResponse> {
    return this.client.request(
      'DELETE',
      `/api/v1/external-agents/tasks/${taskId}`,
    ) as Promise<CancelResponse>;
  }
}
